plugins/woocommerce/src/Internal/StockNotifications/Notification.php

```diff
@@ -449,4 +449,117 @@ public function get_product_name() {
 
 		return $product->get_parent_id() ? $product->get_name() : $product->get_title();
 	}
+
+	/**
+	 * Check if the notification has valid verification data.
+	 *
+	 * This is used to determine if the notification can be verified via email link.
+	 *
+	 * @return bool
+	 */
+	public function has_valid_verification_data(): bool {
+		$action_key = $this->get_meta( 'email_link_action_key' );
+
+		if ( empty( $action_key) ) {
+			return false;
+		}
+
+		if ( ! str_contains( $action_key, ':' ) ) {
+			return false;
+		}
+
+		list( $requested_timestamp, $key ) = explode( ':', $action_key, 2 );
+		$expiration_duration = Config::get_unverified_deletion_days_threshold() * DAY_IN_SECONDS;
+
+		if ( 0 !== $expiration_duration && ( time() - (int) $requested_timestamp ) > $expiration_duration ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Setup verification data for the notification.
+	 *
+	 * This is used to generate a unique key for email verification links.
+	 */
+	public function setup_verification_data() {
+		$key = time() . ':' . wp_generate_password( 20, false );
+		$this->update_meta_data( 'email_link_action_key', $key );
+	}
+
+	/**
+	 * Maybe setup verification data for the notification.
+	 *
+	 * This is used to ensure that the notification has valid verification data.
+	 *
+	 * @param bool $persist Whether to persist the changes to the database.
+	 */
+	public function maybe_setup_verification_data( bool $persist ): void {
+		$opt_in_required = get_option( 'woocommerce_customer_stock_notifications_opt_in_required', 'no' );
```

Comment:
I believe this isn't necessary and, imo, slightly blurs the separation of concerns.